core/src/main/scala/kafka/server/MetadataCache.scala

```diff
@@ -42,63 +45,86 @@ private[server] class MetadataCache(brokerId: Int) extends Logging {
 
   this.logIdent = "[Kafka Metadata Cache on broker %d] ".format(brokerId)
 
-  def getTopicMetadata(topics: Set[String], protocol: SecurityProtocol) = {
+  private def getAliveEndpoints(brokers: Iterable[Int], protocol: SecurityProtocol): Seq[BrokerEndPoint] = {
+    brokers.map(aliveBrokers.getOrElse(_, null)).filter(_ != null).toSeq.map(_.getBrokerEndPoint(protocol))
```

Comment:
Since we now know that this is a hotspot in this code, we should reduce the number of traversals. Relevant code from the trunk commit:

https://github.com/apache/kafka/commit/764d8ca9eb0aba6099ba289a10f437e72b53ffec#diff-bfeebf48d90e86c1ffb850fbbe019dd6R49